src/Compilers/CSharp/Portable/Symbols/Metadata/PE/NativeIntegerTypeDecoder.cs

```diff
@@ -27,24 +34,42 @@ private static TypeSymbol TransformType(TypeSymbol type, ImmutableArray<bool> tr
             try
             {
                 var result = decoder.TransformType(type);
+                Debug.Assert(decoder._decodeStatus == DecodeStatus.NotFailed);
                 if (decoder._index == transformFlags.Length)
                 {
                     return result;
                 }
+                else
+                {
+                    return new UnsupportedMetadataTypeSymbol();
+                }
             }
             catch (ArgumentException)
             {
+                if (decoder._decodeStatus == DecodeStatus.FailedToErrorType)
+                {
+                    // If we failed to decode because there was an error type involved, marking the
+                    // metadata as unsupported means that we'll cover up the error that would otherwise
+                    // be reported for the type. This would likely lead to a worse error message as we
+                    // would just report a BindToBogus, so return the type unchanged.
+                    Debug.Assert(type.ContainsErrorType());
+                    return type;
+                }
+
+                Debug.Assert(decoder._decodeStatus == DecodeStatus.FailedToBadMetadata);
+                return new UnsupportedMetadataTypeSymbol();
             }
-            return type;
         }
 
         private readonly ImmutableArray<bool> _transformFlags;
         private int _index;
+        private DecodeStatus _decodeStatus;
```

Comment:
>DecodeStatus _decodeStatus [](start = 16, length = 26)

Is this field needed?

Rather than using a field, could we just throw `ArgumentException` for cases that have error types (where we catch the exception locally and return the original type), and throw `UnsupportMetadataTypeSymbol` for case where we fail completely?